packages/core/parcel-bundler/src/assets/JSAsset.js

```diff
@@ -153,17 +153,41 @@ class JSAsset extends Asset {
     await babel(this);
 
     // Inline environment variables
-    if (this.options.target === 'browser' && ENV_RE.test(this.contents)) {
+    const hasProcessEnv = ENV_RE.test(this.contents);
+    if (this.options.target === 'browser' && hasProcessEnv) {
       await this.parseIfNeeded();
       this.traverseFast(envVisitor);
     }
 
     // Inline process.browser
-    if (this.options.target === 'browser' && BROWSER_RE.test(this.contents)) {
+    const hasProcessBrowser = BROWSER_RE.test(this.contents);
+    if (this.options.target === 'browser' && hasProcessBrowser) {
       await this.parseIfNeeded();
       this.traverse(processVisitor);
       this.isAstDirty = true;
     }
+
+    if (
+      this.options.minify &&
+      this.options.target === 'browser' &&
+      (hasProcessBrowser || hasProcessEnv)
+    ) {
+      await babel7(this, {
+        internal: true,
+        config: {
+          plugins: [
+            [
+              require('babel-plugin-minify-dead-code-elimination'),
```

Comment:
I am actually not sure about the library. As we already have terser to eliminate dead code, we might better use it here as well, but as it's not easy to make it compatible with the babel ast this was the easiest solution I found. Maybe you have some thoughts on this and can show me how to use terser here.

Reply:
Terser already runs later so I'm not sure why you would need this.

Reply:
If you disable this lines of code, the test will fail. The issue is, that collectDependencies runs after the pretransform. Therefore the dependencies are in the bundle. The code within the falsy if case will be removed by terser in the end, but the collected dependencies will stay. Maybe something else is wrong here, but with this lines of code I saved ~200kb of JavaScript for my application. Maybe you have a better idea how to solve this?
I guess this will be solved by scope hoisting as well, but as it's experimental for know it would be nice to solve this for now.

Reply:
@devongovett If unused branches (after the env and process.browser visitors) contain requires, these requires will actually be part of the bundle.

@garthenweb A cleaner (and faster) solution is to change the env and process visitor to automatically remove unreachable branches if possible. A start:
```diff
diff --git a/packages/core/parcel-bundler/src/visitors/env.js b/packages/core/parcel-bundler/src/visitors/env.js
index 295c654c..468f2d21 100644
--- a/packages/core/parcel-bundler/src/visitors/env.js
+++ b/packages/core/parcel-bundler/src/visitors/env.js
@@ -1,7 +1,8 @@
 const types = require('@babel/types');
 
 module.exports = {
-  MemberExpression(node, asset) {
+  MemberExpression(path, asset) {
+    const {node} = path;
     // Inline environment variables accessed on process.env
     if (types.matchesPattern(node.object, 'process.env')) {
       let key = types.toComputedKey(node);
@@ -12,6 +13,24 @@ module.exports = {
           morph(node, value);
           asset.isAstDirty = true;
           asset.cacheData.env[key.value] = process.env[key.value];
+          if (types.isUnaryExpression(path.parent, {operator: '!'})) {
+            path = path.parentPath;
+          }
+          if (
+            types.isConditionalExpression(path.parent) ||
+            types.isIfStatement(path.parent)
+          ) {
+            const result = path.evaluate();
+            if (result.confident) {
+              if (result.value && path.parent.consequent) {
+                path.parentPath.replaceWith(path.parent.consequent);
+              } else if (!result.value && path.parent.alternate) {
+                path.parentPath.replaceWith(path.parent.alternate);
+              } else {
+                path.parentPath.remove();
+              }
+            }
+          }
         }
       }
     }

diff --git a/packages/core/parcel-bundler/src/visitors/process.js b/packages/core/parcel-bundler/src/visitors/process.js
index 4a3c209b..8e102bb0 100644
--- a/packages/core/parcel-bundler/src/visitors/process.js
+++ b/packages/core/parcel-bundler/src/visitors/process.js
@@ -5,12 +5,27 @@ module.exports = {
     // Inline process.browser
     const isProcess = path.node.object.name === 'process';
     const isBrowser = path.node.property.name === 'browser';
-    const isAssignment = path.parentPath.type === 'AssignmentExpression';
     if (isProcess && isBrowser) {
-      if (isAssignment) {
+      if (t.isAssignmentExpression(path.parentPath)) {
         path.parentPath.remove();
       } else {
-        path.replaceWith(t.booleanLiteral(true));
+        if (t.isIfStatement(path.parentPath)) {
+          path.parentPath.replaceWith(path.parent.consequent);
+        } else if (
+          t.isUnaryExpression(path.parentPath) &&
+          path.parent.operator === '!' &&
+          t.isIfStatement(path.parentPath.parentPath)
+        ) {
+          if (path.parentPath.parent.alternate) {
+            path.parentPath.parentPath.replaceWith(
+              path.parentPath.parent.alternate
+            );
+          } else {
+            path.parentPath.parentPath.remove();
+          }
+        } else {
+          path.replaceWith(t.booleanLiteral(true));
+        }
       }
     }
   }     
diff --git a/packages/core/parcel-bundler/src/assets/JSAsset.js b/packages/core/parcel-bundler/src/assets/JSAsset.js
index 5d42aca0..48024ae3 100644
--- a/packages/core/parcel-bundler/src/assets/JSAsset.js
+++ b/packages/core/parcel-bundler/src/assets/JSAsset.js
     // Inline environment variables
     if (this.options.target === 'browser' && ENV_RE.test(this.contents)) {
       await this.parseIfNeeded();
-      this.traverseFast(envVisitor);
+      this.traverse(envVisitor);
     }
```

Reply:
Thank you for the advice, I will check as soon as I find time for it.

Reply:
@mischnic I implemented your feedback now and build a stand-alone version to remove the dead code.
The failing tests don't look related to my changes, an earlier build succeeded, see https://dev.azure.com/devongovett/devongovett/_build/results?buildId=1363. I only pushed small cleanup changes afterward.
I had the feeling that it makes more sense to move the `process.js` and `env.js` file together as they are doing more or less the same. Let me know if you have any concerns about this.

As always, I am happy to receive feedback!